src/hermes/model/context_manager.py

```diff
@@ -55,10 +55,19 @@ def prepare_step(self, step: str, *depends: str) -> None:
         self._current_step.append(step)
 
     def finalize_step(self, step: str) -> None:
-        current_step = self._current_step.pop()
+        if len(self._current_step) < 1:
+            raise ValueError("There is no step to end.")
+        current_step = self._current_step[-1]
         if current_step != step:
```

Comment:
```suggestion
        if self._current_step[-1] != step:
```

No need to keep the local variable... or use it in the f-string in 62, too.